sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/MutationKeyEncoder.java

```diff
@@ -77,7 +77,7 @@ private void encodeKey(OrderedCode orderedCode, Mutation m) {
     Map<String, Value> mutationMap = mutationAsMap(m);
     for (SpannerSchema.KeyPart part : schema.getKeyParts(m.getTable())) {
       Value val = mutationMap.get(part.getField());
-      if (val.isNull()) {
+      if (val == null || val.isNull()) {
```

Comment:
Consider the behavior of a normal insert statement that omits a column value:

```sql
CREATE TABLE test (
    key1 String(MAX), 
    key2 STRING(max)) 
    PRIMARY KEY (key1, key2)
INSERT INTO test 
    (key1) 
    VALUES ("a")
```
This will result in a row with values `("a", <NULL>)`
So assuming a missed value == NULL is the same behavior as the database.

Not that this encoding is only used to sort mutations by table and key -- there is no equivalent decoder and values encoded in this way are never written to the DB. 

NOT NULL columns can appear anywhere, not just in keys, and will cause the insert to fail at a later stage, at which point the mutation will be added to the FailedMutationGroups PCollection which is presented as the output of SpannerIO.Write

Reply:
The database engine considers missing values as NULL - this code is mimicking the same behavior. 

Note also that this code is only concerned with Primary Key columns, and the encoded value is only used for sorting mutations in a bundle by table and primary key -- the value encoded here is never written to the database.

NOT NULL constraints can be set on any column. If the Mutation is an insert and does not specify a column which is NOT NULL (or sets it to NULL), then the insert will fail, and the Mutation appended to a PCollection of failed MutationGroups which is output by the SpannerIO.Write transform.

Reply:
Ok. Thanks for clarifying.